File: typegraph/python_next/typegraph_next/runtimes/deno.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, List, Optional

from typegraph_next.gen.exports.runtimes import (
    Effect,
    EffectNone,
    MaterializerDenoFunc,
    MaterializerDenoImport,
    MaterializerDenoPredefined,
)
from typegraph_next.gen.types import Err
from typegraph_next.policy import Policy
from typegraph_next.wit import runtimes, store

if TYPE_CHECKING:
    from typegraph_next import t


@dataclass
class Runtime:
    id: int


@dataclass
class Materializer:
    id: int


class DenoRuntime(Runtime):
    def __init__(self):
        super().__init__(runtimes.get_deno_runtime(store))

    def func(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        code: str,
        secrets: Optional[List[str]] = None,
        effect: Optional[Effect] = None,
    ):
        secrets = secrets or []
        effect = effect or EffectNone()
        mat_id = runtimes.register_deno_func(
            store,
            MaterializerDenoFunc(code=code, secrets=secrets),
            effect,
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        from typegraph_next import t

        return t.func(
            inp, out, FunMat(mat_id.value, code=code, secrets=secrets, effect=effect)
        )

    def import_(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        module: str,
        name: str,
        effect: Optional[Effect] = None,
        secrets: Optional[List[str]] = None,
    ):
        effect = effect or EffectNone()
        secrets = secrets or []
        mat_id = runtimes.import_deno_function(
            store,
            MaterializerDenoImport(func_name=name, module=module, secrets=secrets),
            effect,
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        from typegraph_next import t

        return t.func(
            inp,
            out,
            ImportMat(
                id=mat_id.value,
                name=name,
                module=module,
                secrets=secrets,
                effect=effect,
            ),
        )

    def identity(self, inp: "t.struct") -> "t.func":
        from typegraph_next import t

        res = runtimes.get_predefined_deno_func(
            store, MaterializerDenoPredefined(name="identity")
        )
        if isinstance(res, Err):
            raise Exception(res.value)

        return t.func(
            inp,
            inp,
            PredefinedFunMat(id=res.value, name="identity"),
        )

    def policy(
        self, name: str, code: str, secrets: Optional[List[str]] = None
    ) -> Policy:
        secrets = secrets or []
        mat_id = runtimes.register_deno_func(
            store,
            MaterializerDenoFunc(code=code, secrets=secrets),
            EffectNone(),
        )

        if isinstance(mat_id, Err):
            raise Exception(mat_id.value)

        return Policy.create(
            name,
            mat_id.value,
        )

    def import_policy(
        self,
        func_name: str,
        module: str,
        secrets: Optional[List[str]] = None,
        name: Optional[str] = None,
    ) -> Policy:
        name = name or re.sub("[^a-zA-Z0-9_]", "_", f"__imp_{module}_{name}")

        res = runtimes.import_deno_function(
            store,
            MaterializerDenoImport(
                func_name=func_name, module=module, secrets=secrets or []
            ),
            EffectNone(),
        )
        if isinstance(res, Err):
            raise Exception(res.value)

        return Policy.create(name, res.value)


@dataclass
class FunMat(Materializer):
    code: str
    secrets: List[str]
    effect: Effect


@dataclass
class ImportMat(Materializer):
    module: str
    name: str
    effect: Effect
    secrets: List[str]


@dataclass
class PredefinedFunMat(Materializer):
    name: str

File: typegraph/node/src/runtimes/deno.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.js";
import { runtimes } from "../wit.js";
import { Effect } from "../gen/interfaces/metatype-typegraph-runtimes.js";
import Policy from "../policy.js";
import { Materializer, Runtime } from "./mod.js";
import { fx } from "../mod.js";

interface FunMat extends Materializer {
  code: string;
  secrets: string[];
  effect: Effect;
}

interface ImportMat extends Materializer {
  name: string;
  module: string;
  secrets: string[];
  effect: Effect;
}

interface PredefinedFuncMat extends Materializer {
  name: string;
}

export interface DenoFunc {
  code: string;
  secrets?: Array<string>;
  effect?: Effect;
}

export interface DenoImport {
  name: string;
  module: string;
  secrets?: Array<string>;
  effect?: Effect;
}

export class DenoRuntime extends Runtime {
  constructor() {
    super(runtimes.getDenoRuntime());
  }

  func<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(
    inp: I,
    out: O,
    { code, secrets = [], effect = fx.read() }: DenoFunc,
  ): t.Func<P, I, O, FunMat> {
    const matId = runtimes.registerDenoFunc({ code, secrets }, effect);
    const mat: FunMat = {
      _id: matId,
      code,
      secrets,
      effect,
    };
    return t.func(inp, out, mat);
  }

  import<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
    O extends t.Typedef = t.Typedef,
  >(
    inp: I,
    out: O,
    { name, module, effect = fx.read(), secrets = [] }: DenoImport,
  ): t.Func<P, I, O, ImportMat> {
    const matId = runtimes.importDenoFunction({
      funcName: name,
      module,
      secrets,
    }, effect);
    const mat: ImportMat = {
      _id: matId,
      name,
      module,
      secrets,
      effect,
    };
    return t.func(inp, out, mat);
  }

  identity<
    P extends Record<string, t.Typedef> = Record<string, t.Typedef>,
    I extends t.Struct<P> = t.Struct<P>,
  >(inp: I): t.Func<P, I, I, PredefinedFuncMat> {
    const mat: PredefinedFuncMat = {
      _id: runtimes.getPredefinedDenoFunc({ name: "identity" }),
      name: "identity",
    };
    return t.func(
      inp,
      inp,
      mat,
    );
  }

  policy(name: string, _code: string): Policy;
  policy(name: string, data: Omit<DenoFunc, "effect">): Policy;
  policy(name: string, data: string | Omit<DenoFunc, "effect">): Policy {
    const params = typeof data === "string"
      ? { code: data, secrets: [] }
      : { secrets: [], ...data };

    return Policy.create(
      name,
      runtimes.registerDenoFunc(params, fx.read()),
    );
  }

  importPolicy(data: Omit<DenoImport, "effect">, name?: string): Policy {
    const policyName = name ?? `__imp_${data.module}_${data.name}`.replace(
      /[^a-zA-Z0-9_]/g,
      "_",
    );
    return Policy.create(
      policyName,
      runtimes.importDenoFunction({
        funcName: data.name,
        module: data.module,
        secrets: data.secrets ?? [],
      }, fx.read()),
    );
  }
}
